govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/context_relevancy/context_relevancy.py

```diff
@@ -0,0 +1,202 @@
+                    f"Verdicts:\n{prettify_list(verdict_collection.verdicts)}",
+                    f"Score: {self.score}\nReason: {self.reason}",
+                ],
+            )
+
+            return self.score
+
+    async def _generate_truths(self, retrieval_context: list[str]) -> TruthCollection:
+        prompt = self.evaluation_template.truths(
+            retrieval_context=retrieval_context,
+        )
+        return await self._generate_result_from_model(prompt, schema=TruthCollection)
+
+    async def _generate_information_needs(
+        self, input: str
+    ) -> InformationNeedsCollection:
+        prompt = self.evaluation_template.information_needs(input=input)
+        return await self._generate_result_from_model(
+            prompt, schema=InformationNeedsCollection
+        )
+
+    async def _generate_verdicts(
+        self,
+        information_needs_collection: InformationNeedsCollection,
+        truth_collection: TruthCollection,
+    ) -> VerdictCollection:
+        if len(information_needs_collection.information_needs) == 0:
+            return VerdictCollection(verdicts=[])
+
+        extracted_truths = [truth.model_dump() for truth in truth_collection.truths]
+        prompt = self.evaluation_template.verdicts(
+            information_needs=information_needs_collection.information_needs,
+            extracted_truths=extracted_truths,
+        )
+
+        return await self._generate_result_from_model(prompt, schema=VerdictCollection)
+
+    def _calculate_score(self, verdicts: VerdictCollection) -> float:
+        score = verdicts.score_verdicts()
+
+        return 0 if self.strict_mode and score < self.threshold else score
+
+    async def _generate_reason(
+        self, input: str, verdict_collection: VerdictCollection
+    ) -> str | None:
+        if self.include_reason is False:
+            return None
+
+        unmet_needs = verdict_collection.unmet_needs()
+        score = float(round(self.score or 0.0, 2))
+
+        prompt = self.evaluation_template.reason(
+            unmet_needs=unmet_needs,
+            input=input,
+            score=score,
+        )
+
+        result = await self._generate_result_from_model(prompt, schema=ScoreReason)
+        return result.reason if isinstance(result.reason, str) else str(result.reason)
+
+    async def _generate_result_from_model(
+        self, prompt: str, schema: Type[SchemaType]
+    ) -> SchemaType:
+        if self.using_native_model:
+            result, cost = await self.model.a_generate(prompt, schema=schema)
+            self._increment_cost(cost)
+
+            # The return sig of a_generate is Tuple[Union[str, Dict], float], but
+            # amazon_bedrock_model.a_generate returns the schema instance directly when a
+            # schema is provided https://github.com/confident-ai/deepeval/blob/main/deepeval/models/llms/amazon_bedrock_model.py#L95-L96
```

Comment:
Thanks for the comment, I think we might actually have the type signature wrong based on a chat on Chae's PR: https://github.com/alphagov/govuk-chat-evaluation/pull/133/files#diff-532ce599a77962860a9082458c0a1d483c61b74fa1a42699b7ce5cb829bb32cd

Where he had a good suggestion of handling it with: https://github.com/alphagov/govuk-chat-evaluation/pull/133/files#diff-96984f29d34bdd4fbc731c0f7681cd2df96e4b8ef96b17cb0f2d68220a8172daR107-R110